File: JavaScript/Easy/ValidParentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  let stack = [];
  const openers = {
    "(": ")",
    "[": "]",
    "{": "}"
  };
  const closers = new Set([")", "}", "]"]);

  for (let i = 0; i < s.length; i++) {
    if (openers[s[i]]) {
      stack.push(s[i]);
    } else if (closers.has(s[i])) {
      if (s[i] !== openers[stack.pop()]) {
        return false;
      }
    }
  }

  return stack.length === 0;
};

File: Python/Easy/ValidParentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) & 1:
            return False

        stack = []

        for char in s:
            if char == "(":
                stack.append(")")
            elif char == "{":
                stack.append("}")
            elif char == "[":
                stack.append("]")

            elif not stack or stack.pop() != char:
                return False

        return not stack
